### PR TITLE
Bring some GL ES 3.0 functions to GetProcAddress

### PR DESCRIPTION
**Pull request by kvark**

### CODE REVIEW

File: system/lib/gl.c [hunk 1]
```diff
@@ -1560,174 +1560,240 @@ void* emscripten_GetProcAddress(const char *name_) {
...
+  if (!strcmp(name, "glDrawBuffers")) return emscripten_glDrawBuffers;
+  // UBO
+  if (!strcmp(name, "glGetUniformIndices")) return emscripten_glGetUniformIndices;
+  if (!strcmp(name, "glGetActiveUniformsiv")) return emscripten_glGetActiveUniformsiv;
+  if (!strcmp(name, "glGetActiveUniformName")) return emscripten_glGetActiveUniformName;
+  if (!strcmp(name, "glGetUniformBlockIndex")) return emscripten_glGetUniformBlockIndex;
+  if (!strcmp(name, "glGetActiveUniformBlockiv")) return emscripten_glGetActiveUniformBlockiv;
+  if (!strcmp(name, "glGetActiveUniformBlockName")) return emscripten_glGetActiveUniformBlockName;
+  if (!strcmp(name, "glUniformBlockBinding")) return emscripten_glUniformBlockBinding;
+  // GLES 3.0 functions
```
Comment (kripken) [inline system/lib/gl.c]: I assume all the new lines are from here, and above this it was just removing the `else`? (which I agree is a nice change)

Comment (kvark) [reply system/lib/gl.c]: yes, exactly

**Comment by kvark** [comment]
Uh oh, noticed the following comment explaining the issue (in docs/optimizing/Optimizing-WebGL.rst):
> Avoid using any of the ``*glGetProcAddress()`` API functions. Emscripten provides static linking to all of the GL API functions, even for all WebGL extensions. The ``*glGetProcAddress()`` API is only provided for compatibility to ease porting of existing code, but accessing WebGL via calling dynamically obtained function pointers is noticeably slower than direct function calls, due to extra function pointer security validation that dynamic dispatching has to do in asm.js/WebAssembly. Since Emscripten provides all of the GL entry points statically linked in, it is recommended to take advantage of this for best performance.

**Comment by kvark** [comment]
@kripken this behavior seems to contradict EGL specification:
> eglGetProcAddress may be queried for all EGL and client API functions supported by the implementation (whether those functions are extensions or not, and whether they are supported by the current client API context or not).
> For functions that are queryable with eglGetProcAddress,  implementations may choose to also export those functions statically from the object libraries implementing those functions. However, portable clients cannot rely on this behavior.

:(

**Comment by kripken** [comment]
I don't understand, what's the contradiction?

**Comment by kvark** [comment]
@kripken if Emscripten is implementing context creation via EGL, it has to obey the EGL specification, which states that `GetProcAddress` should always give you the function pointers. In this case, the documentation paragraph I quoted is incorrect. If it is correct though, then my PR is not valid. So this is waiting for your confirmation on which side is right (EGL spec + my PR, or the current Emscripten docs).

Also, I found a largely similar (but better) work - #4829

**Comment by kripken** [comment]
@juj would know best what to do here. Also, the status of that other PR.

But I don't see a problem in the documentation paragraph - it just says the function pointers are slower, it doesn't say if they will or won't be supported? They should be supported, I think. Maybe we can improve the text there, but I'm not sure how?

**Comment by juj** [comment]
This PR looks good to me (will need rebasing to resolve merge conflict). We want to support both forms - querying via function pointers and static linking, and we encourage everyone to use static linking, because we have that opportunity. Static linking is faster in practice - invoking via dynamic pointers does show up in profiles as a few % unnecessary overhead, plus also `*GetProcAddress()` functions have a horrible dead code elimination issue: referencing a `*GetProcAddress()` function will cause all GL/EGL functions to be pulled in to the build since there is no machinery that would know which one of the functions will be queried at runtime, so all functions will need to be present. This is really bad for minimizing code size, so it's best for people not to rely on `*GetProcAddress()` to get the best code size.

Still, we do want to support `*GetProcAddress()` properly for good form and porting experience, so it is made available if necessary.

**Comment by kvark** [comment]
@juj thanks for clarifying. I rebased the PR now.
It's still lacking testing and full coverage of exported functions.

**Comment by kripken** [comment]
Hi @kvark - sorry this PR got forgotten!

I did some work to apply this over current incoming, here are two commits that should do that, first does the else removal, second adds your new functions here:

https://github.com/kripken/emscripten/commit/0dad3879d28ccab5001ddd51ce5112783b052674

https://github.com/kripken/emscripten/commit/953c3bf6959540916b15d3873f7fb9be36ce072f

I think you can start with current incoming, cherry-pick those two commits, and add yourself to AUTHORS, and then we can merge this.

**Comment by kvark** [comment]
@kripken please take another look

**Comment by kripken** [comment]
Great, thanks @kvark !

**Pull request merged**